### PR TITLE
chapter_10

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -165,3 +165,15 @@ input {
   margin-left: 0;
   width: auto;
 }
+
+/* Users index */
+
+.users {
+  list-style: none;
+  margin: 0;
+  li {
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: add an empty line

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -6,8 +6,10 @@ def create
     if user&.authenticate params[:session][:password]
       log_in user
       params[:session][:remember_me].eql? Settings.source.collection.session_controller_create ? remember(user) : forget(user)
-      redirect_to user
+      redirect_back_or user
     else
+      message = "Account not activated. "
+      message += "Check your email for the activation link."
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]: using I18n for this text

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,32 +1,77 @@
 class UsersController < ApplicationController
-  before_action :userprofile, only: :show
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
+  before_action :load_user, only: [:show, :edit, :update]
+  before_action :correct_user, only: [:edit, :update]
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: image

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,32 +1,77 @@
...
-      flash.now[:danger] = "account cannot be created"
+      flash.now[:danger] = t ".fail_to_create"
       render :new
     end
   end 
 
+  def edit; end
+
+  def update
+    if @user.update(user_params)
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,32 +1,77 @@
...
+      flash[:success] =  t ".updated_sucessfully"
+    else
+      flash.now[:danger] = t ".cannot_be_updated"
+      render :edit
+    end
+  end
+
+  def destroy
+    User.find(params[:id]).destroy
+    flash[:success] = "User deleted"
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
    if @user.destroy
      flash[:success] = "User deleted"
    else 
      flash[:danger] = "Error"
```

File: app/views/users/_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<%= form_for(@user) do |f| %>
```
Comment (saobv-2223) [inline app/views/users/_form.html.erb]: remove ()

File: app/views/users/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<% provide :title, "Edit user" %>
+<% provide(:button_text, 'Save changes') %>
```
Comment (saobv-2223) [inline app/views/users/edit.html.erb]: remove () and using "

File: app/views/users/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+<h1>Update your profile</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= render "form" %>
+    <div class="gravatar_edit">
+      <%= gravatar_for @user %>
+      <a href="https://gravatar.com/emails" target="_blank">Change</a>
```
Comment (saobv-2223) [inline app/views/users/edit.html.erb]: using link_to

**Comment by chuan97sun** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 9]
```diff
@@ -165,3 +165,16 @@ input {
...
 }
+
+/* Users index */
+
+.users {
+  list-style: none;
+  margin: 0; 
+  li {
+  
+    overflow: auto;
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]:
```suggestion
  margin: 0; 
  
  li {
    overflow: auto;
```

File: app/controllers/sessions_controller.rb [hunk 10]
```diff
@@ -6,9 +6,11 @@ def create
     if user&.authenticate params[:session][:password]
       log_in user
       params[:session][:remember_me].eql? Settings.source.collection.session_controller_create ? remember(user) : forget(user)
-      redirect_to user
+      redirect_back_or user
     else
-      flash.now[:danger] = "Invalid email/password combination" 
+      message = t ".message_one"
+      message += t ".message_two"
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]:
```suggestion
```

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -1,32 +1,79 @@
 class UsersController < ApplicationController
-  before_action :userprofile, only: :show
+  before_action :logged_in_user, only: %i(index edit update destroy)
+  before_action :load_user, only: %i(edit update destroy show)
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: image

Comment (chuan97sun) [reply app/controllers/users_controller.rb]: I cannot load the sign in page and log in  error message after i change only to except

**Comment by chuan97sun** [comment]
ready

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -1,32 +1,79 @@
...
   def create
     @user = User.new user_params
     if @user.save
       log_in @user
-      flash[:success] = "Welcome to the Sample App!"
+      flash[:success] = t(".successfully_create")
       redirect_to @user  
     else
-      flash.now[:danger] = "account cannot be created"
+      flash.now[:danger] = t(".fail_to_create")
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -1,32 +1,79 @@
...
-     redirect_to root_path
-  end
+  def show; end
 
   def create
     @user = User.new user_params
     if @user.save
       log_in @user
-      flash[:success] = "Welcome to the Sample App!"
+      flash[:success] = t(".successfully_create")
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -1,32 +1,79 @@
...
   end 
 
+  def edit; end
+
+  def update
+    if @user.update user_params
+      redirect_to @user
+      flash[:success] =  t(".updated_sucessfully")
+    else
+      flash.now[:danger] = t(".cannot_be_updated")
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -1,32 +1,79 @@
...
+      render :edit
+    end
+  end
+
+  def destroy
+     if @user.destroy
+      flash[:success] =  t(".User_deleted")
+    else
+      flash[:warning] = t(".fail_to_delete")
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -1,32 +1,79 @@
...
 
   def user_params
     params.require(:user).permit User::USERS_PARAMS
   end
+
+  def load_user
+     @user  = User.find_by id: params[:id]
+     return if @user
+     flash[:warning] = t(".user_cannot_be_detected")
+     redirect_to root_path
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
     return if @user
     
     flash[:warning] = t ".user_cannot_be_detected"
     redirect_to root_path
```

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -1,32 +1,79 @@
...
+  end
+
+  def logged_in_user
+    unless logged_in?
+      store_location
+      flash[:danger] = t (".Please_log_in")
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove ()

File: app/helpers/sessions_helper.rb [hunk 18]
```diff
@@ -36,4 +36,17 @@ def log_out
     session.delete :user_id
     @current_user = nil
   end
+
+  def current_user? user
+    user && user == current_user 
+  end
+   
+  def redirect_back_or default
+    redirect_to(session[:forwarding_url] || default)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

**Comment by chuan97sun** [comment]
ready

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**